### ISSUE TITLE
Applications should not auto-allocate bricks

### ISSUE THREAD

**Issue by dwf** [mila-udem/blocks#936]
Related to #935 but it deserves its own discussion.

In order to better respect [command query separation](https://en.wikipedia.org/wiki/Command%E2%80%93query_separation) (although technically apply methods are more like factories than queries), I'd like to see what people think of requiring explicit allocation and/or initialization before applications get called. I'd argue that the current behaviour is not an implied part of a user's mental model (even though I knew deep down it would happen, I didn't think of this as a source of subtle bugs).

TL;DR Initialization can still implicitly allocate, but `apply()` before allocation (explicit or implicit) should raise an error. Thoughts?

**Comment by dwf** [comment]
Just occurred to me how bloody annoying this would be for bricks that require no allocation, i.e. it would make `x = tensor.matrix(); y = Logistic().apply(x)` no longer work. So any solution would need to be more nuanced.
